aten/src/ATen/native/ReduceAllOps.cpp

```diff
@@ -34,6 +34,19 @@ Tensor max(const Tensor &self) {
 }
 
 Tensor& max_unary_out(const Tensor &self, Tensor& out) {
+
+  // First check if the devices match (CPU vs GPU)
+  TORCH_CHECK(self.device() == out.device());
+
+  TORCH_CHECK(canCast(
+      typeMetaToScalarType(self.dtype()),
+      typeMetaToScalarType(out.dtype())));
+
+  if (self.numel() == 0) {
+    at::native::resize_output(out, self.sizes());
+    return out;
```

Comment:
This is implemented pretty inefficiently - the out= kernel just allocates a temporary by calling `at::max()`.

I think it shouldn't be too hard to implement the efficient way: the existing `max()` kernel in this file just allocates an empty tensor and calls `max_all_stub`, which already takes an out= tensor.

The easiest way to update that would be to update this kernel to do the same (although that means this kernel will only be able to run on cpu and cuda. `max()` also has an MPS and QuantizedCPU`, so if you want this new op to support those backends you'd need to write the efficient kernel for them too.)

Reply:
Thanks for the response! I will take a look. If `max()` calls  `max_all_stub`, why would our change not compatible with other backends? I am assuming `max.unary_out()` would probably also just call `max_all_stub`?

Reply:
From separate chat - it's probably enough to make the fast kernel for cpu/cuda, and the slow kernel for now for the other backends (like QuantizedCPU)